fix problem with long Url.pattern from model_bakery and admin

PetrDlouhy opened the pull request:
fix #132 

- The `MaxLengthValidator` has problem, that it doesn't allow `/` with 254 other characters (which is generated when saved previously with long pattern). But I suppose it is very minor problem not worth adding more complexity to the code.

tony commented:
Thank you for this!

At this time, fixing #124 is a more urgent issue. I believe that will need to get resolved first.

I also note there's no test - which we probably want - but not having written one is understandable given the project's current state.

tony commented:
I've created #136 to improve test and development infrastructure. 

I will also document the test flow

tony commented:
I haven't merged anything yet, but as-is (5.0 at 77b980c), here's how tests can be ran:

```console
env PYTHONPATH=. DJANGO_SETTINGS_MODULE=tests.settings django-admin test robots -v2
```

PetrDlouhy commented:
I have added test for this Issue and few other Url model tests.

Although the test would raise `django.db.utils.DataError: value too long for type character varying(255)` only on PostgreSQL testing database. SQLite is OK with longer than expected strings.

tony commented:
@PetrDlouhy Sorry for the delay.

Are you still interested in this PR? Can you rebase this?

PetrDlouhy commented:
@tony I rebased this.

tony commented on src/robots/models.py:
```diff
@@ -21,6 +22,8 @@ class Url(models.Model):
             " (*) as a wildcard and a dollar sign ($) to "
             "match the end of the URL, e.g., '/*.jpg$'."
         ),
+        default="",
```
Q: Is this `default=""` necessary?

I am not sure, but couldn't this lock tables / cause potential downtimes on production sites?

PetrDlouhy commented:
@tony I don't rememeber why I added the `default=""`. I got rid of it and it seems, that tests are passing, so I would assume it is not necessary.

tony commented:
@PetrDlouhy First, thank you for your time. ☺️ 

In regards to the description in the PR and in #132. Correct me if I'm mistaken below! 

To summarize this PR: it adds a validator to solve a compatibility issue arising with a third party django package: [`model_bakery`](https://github.com/model-bakers/model_bakery). Does this sound accurate?

Does this problem still effect you? e.g. Do you still run model_bakery and django-robots and run into this error? Sorry time has passed since then. 

If this issue still happens for you: Is the PR still need needed now that `blank=''` was removed in [8e5d9d9](https://github.com/jazzband/django-robots/pull/133/commits/8e5d9d99c4f57a6b8f94003a6b0c283bfe3a3bd2)? How do we know it's a bug with django-robots and not resolvable via `model_bakery`?

FYI: I haven't used model bakery FYI, so things are not obvious to me.

Thank you!

PetrDlouhy commented:
@tony I don't think this is an issue only for `model_bakery`, but in that usage it becomes most prominent.

The problem is, that if you write 255 characters into the `URL.pattern` field it will end up in exception. The reason is that, the `save()` method will add the leading slash (`/`) character making the string 256 characters long.
The error does arise as well when you do this in `django-admin`. Instead of convenient validation error the user would get server error 500.

The problem with `model_bakery` is that it would crate the `pattern` string always 255 characters long, so that this error will arise anytime I would like to generate an `URL`.

And yes, I am still using `model_bakery` for this.

tony commented:
@PetrDlouhy I am grateful for the pull request, thank you. Sorry for the delay!

I am concerned about merging in something that'd potentially have unintended side effects for users. 

## Questions

While the underlying change makes sense to me, no other user has brought up a concern yet since #132 was made. To help reframe this toward django / robots usage in general:

1. Are there any generic use cases to this change beyond third party [model bakery](https://github.com/model-bakers/model_bakery) package?
2. Is there a way in [model bakery](https://github.com/model-bakers/model_bakery) to workaround this issue?
3. In django-robots, is there an alternative approach to accomplish this without changing the model and introducing migration? 

   - If so, what makes this the best choice (in your perspective)?
   - What do we have to assure us this won't break other usages downstream?
   
      As an example:
      - we have our existing tests and the tests you added in the PR
      - this usage is considered canonical django, e.g. django model validators and such are established practice.
      - if merged, we can release the change a pre-release (though I'm not sure if those are in fact used downstream) 

## Second opinion?

Can you please seek a second opinion from someone on Jazzband? The reason why is they would have an opinion that could bring insight and balance out mine.

## Thank you again!

Let me know what you think on both of the above!

PetrDlouhy commented:
@tony The migration is created by Django and is only required because the model state changes but doesn't make any real data change in the DB.
I could not create the migration and everything will work without problems, only the migration checks would throw exception.

I really don't understand your concerns. What exactly could go wrong with this change? What are the unintended side effects?

My use case is that I am using [django-admin-smoke-tests](https://github.com/greyside/django-admin-smoke-tests) for testing if everything works OK in the admin environment.

The problem couldn't be fixed in `model bakery`, because the error is in `django-robots`.

Please go to the Robots URL admin, click on "Add URL" button and fill in long string like `sytowQXhbktgUxdOJHmFxhaLqzxWcsdfsdfsdfsdfsdfsdfqXNbafrVMXCLOjaRELXYSnEzgSgsqzsdfsdfsdfsdfsdfsdfsdfNTqJLkOZPCRvwrqbLtDQmzYghQAaVviHmjoMXRGvXpzJCxNcyfUpJHpaHiBGjyBbdNULZIHGuGUuiyIPkHnTJHNxoVEoDurWjeLgTeqLgSeeZcuzyRkVeNnunAhcaLbOOUwPlXSqqtHORtBLpWIabHLSTcJDIf` and click save.
You will see exception, which is never good.
Users should always see nice pleasant failed validation like this:
![Snímek obrazovky_2023-11-16_16-39-30](https://github.com/jazzband/django-robots/assets/156755/f9d73993-ecac-4d42-9efa-ed0a242fdccf)

Not this:
![Snímek obrazovky_2023-11-16_16-43-06](https://github.com/jazzband/django-robots/assets/156755/85c4f82c-43e6-4ed9-b8c1-721431a6a407)

PetrDlouhy commented:
And yes, model validators are exactly the right tool for this in Django. If the model is not able to accept 255 characters, there should be validator that ensure that users will not put 255 characters there.

tony commented:
@PetrDlouhy 

This looks sensible to me! Pardon the delay.

Let's give this a shot! Merged via 6dbeb311ff674e421e3e724cc71b41afc2aabe45.

P.S. Why the delay? When maintaining, I have to take responsibility for what is merged in: If this affects downstream users in a way that wasn't anticipated, that can inconvenience them. The reason why is an earlier issue happened with migrations in #124.

tony commented:
@PetrDlouhy 

Released 7.0b0 ([Release](https://github.com/jazzband/django-robots/releases/tag/7.0b0), [PyPI](https://pypi.org/project/django-robots/7.0b0/), [Tag](https://github.com/jazzband/django-robots/tree/7.0b0))

If you give 7.0b0 a shot, how is that?

PetrDlouhy commented:
@tony Thank you very much.
I don't know why yet, but tests on my project with 7.0b0 started to fail with `value too long for type character varying(255)`. 
I will have to look at it more closely what makes the difference between my branch and 7.0b0

tony commented:
@PetrDlouhy Thanks! Look forward to you look into this further! You may be in a good position to if you still have `model_bakery`.

PetrDlouhy commented:
@tony I realized, that the tests are not passing because of the removed `default=""` which was in fact fixing `model_bakery` .
I could re-add that and also add some model_bakery tests, if you agree. I don't think it could cause a downtime, because it doesn't change data in the table.

PetrDlouhy commented:
@tony Closing the loop on this, rather late — sorry for leaving it hanging.

**7.0b0 is good.** We have now switched from our fork to `django-robots = "^7.0b0"` from PyPI and it is running fine; the failure I reported above was on our side, not yours.

The mechanism, for the record: `Url.save()` prepends `/` when the pattern lacks one, so `model_bakery` filling `pattern` to `max_length` (255) produces a 256-character value and overflows the column. The `default=""` was masking that, because model_bakery skips fields that have a default — it never generated a pattern at all. So the field was only ever "working" for us by accident.

**I would like to withdraw my offer to re-add `default=""`.** I no longer think it belongs upstream: it changes the library's schema to compensate for a test-tooling artifact in downstream projects, and it hides exactly the overflow that #133 exists to prevent. The right place for the fix is the consumer's test setup — we pin `pattern` when creating `robots.Url` in our admin smoke tests, which is a one-line change and keeps your model honest.

We have also added regression tests around the cap on our side: a pattern over 254 must fail validation, a pattern at exactly 254 must still fit once `save()` prepends the `/` (which is what makes 254 the right number), and the admin must return a form error rather than a `DataError` 500. They fail against 6.1 and pass against 7.0b0.

So from our perspective there is nothing outstanding on #133, and nothing blocking a 7.0 final — which I opened #154 to ask about, before I had re-read this thread. Stable is still 6.1, so everyone not on the prerelease still gets the 500.
